File: src/lib/scala/Slate.Common/src/main/scala/slate/common/Env.scala
/**
<slate_header>
  author: Kishore Reddy
  url: https://github.com/kishorereddy/scala-slate
  copyright: 2015 Kishore Reddy
  license: https://github.com/kishorereddy/scala-slate/blob/master/LICENSE.md
  desc: a scala micro-framework
  usage: Please refer to license on github for more info.
</slate_header>
*/

package slate.common


/**
  * Store the currently selected environment ( local, dev, qa, stg, prod ) and provides some
  * utility functions to parse an environment
  */
object Env extends EnvSupport {
  val PROD = "pro"
  val DEV = "dev"
  val QA  = "qa"
  val UAT = "uat"

  private var _envItem = EnvItem("loc", Env.DEV)


  /**
    * Name of the currently selected environment e.g. ( dev1, qa1, qa2, beta, prod )
    * @return
    */
  def name: String = _envItem.name


  /**
    * Mode of the currently selected environment ( dev, qa, uat, pro )
    * @return
    */
  def mode: String = _envItem.env


  /**
    * The fully qualified name of the currently selected environment ( combines the name + key )
    * @return
    */
  def key: String = _envItem.key


  /**
    * Sets the current environment
    * @param env
    */
  def set(env:EnvItem): Unit =
  {
    _envItem = env
  }


  /**
    * Gets the current environment
    * @return
    */
  def current = _envItem


  /**
    * whether the current environment matches the environment name supplied.
    * @param env
    * @return
    */
  override def isEnv(env: String):Boolean =
  {
    _envItem.env == env
  }


  /**
    * parses the environment name e.g. "qa1:qa" = name:mode
    * @param env
    * @return
    */
  def parse(env:String):EnvItem = {
    val tokens = Strings.split(env, ':')

    // e.g. "dev1", "dev", dev1:dev")
    if(tokens.size == 1){
      return EnvItem(tokens(0), tokens(0))
    }
    EnvItem(tokens(0), tokens(1))
  }
}

File: src/common/envs/library/src/commonMain/kotlin/kiit/common/envs/Env.kt
/**
 *  <kiit_header>
 * url: www.kiit.dev
 * git: www.github.com/slatekit/kiit
 * org: www.codehelix.co
 * author: Kishore Reddy
 * copyright: 2016 CodeHelix Solutions Inc.
 * license: refer to website and/or github
 *  </kiit_header>
 */

package kiit.common.envs

import kotlin.js.JsExport

/**
 * Represents a system environment
 * @param name : e.g. Quality Assurance test
 * @param mode : e.g. Qat
 * @param region : new york
 * @param desc : Qat environment 1 in new york
 */
@JsExport
data class Env(
    val name: String,
    val mode: EnvMode,
    val region: String = "",
    val desc: String = ""
) : EnvSupport {

    /**
     * "qa1:qa"
     * @return
     */
    val key: String get() = name + ":" + mode.name

    override fun isEnvName(envMode: String): Boolean = mode.name == envMode

    override fun isEnv(envMode: EnvMode): Boolean = mode == envMode

    companion object {

        val empty = Env("", EnvMode.Dev, "", "")

        /**
         * parses the environment name e.g. "qa1:qa" = name:mode
         * @param env
         * @return
         */
        fun parse(text: String): Env {
            if (text == "''" || text == "\"\"") {
                return Env("", EnvMode.Dev, desc = "Default environment")
            }
            val tokens = text.split(':')

            // e.g. "dev1", "dev", dev1:dev")
            val env = if (tokens.size == 1)
                Env(tokens[0], interpret(tokens[0]))
            else
                Env(tokens[0], interpret(tokens[1]))

            return env
        }

        /**
         * interprets the string representation of an environment into the type object
         * @param mode
         * @return
         */
        fun interpret(mode: String): EnvMode =
            when (mode) {
                EnvMode.Dev.name -> EnvMode.Dev
                EnvMode.Qat.name -> EnvMode.Qat
                EnvMode.Uat.name -> EnvMode.Uat
                EnvMode.Pro.name -> EnvMode.Pro
                EnvMode.Dis.name -> EnvMode.Dis
                else -> EnvMode.Other(mode)
            }
    }
}
